spirv_hlsl.cpp

```diff
@@ -0,0 +1,1055 @@
+	std::vector<SPIRVariable *> variables;
+	for (auto &id : ids)
+	{
+		if (id.get_type() == TypeVariable)
+		{
+			auto &var = id.get<SPIRVariable>();
+			auto &type = get<SPIRType>(var.basetype);
+
+			if (var.storage != StorageClassFunction && !var.remapped_variable && type.pointer &&
+			    var.storage == StorageClassInput && interface_variable_exists_in_entry_point(var.self))
+			{
+				if (execution.model == ExecutionModelVertex && is_builtin_variable(var))
+					continue;
+				variables.push_back(&var);
+			}
+		}
+	}
+	sort(variables.begin(), variables.end(), VariableComparator(meta));
+	for (auto var : variables)
+	{
+		emit_interface_block_in_struct(*var, binding_number, false);
+	}
+	for (auto var : variables)
+	{
+		emit_interface_block_in_struct(*var, binding_number, true);
+	}
+	end_scope_decl();
+	statement("");
+
+	if (execution.model == ExecutionModelVertex)
+	{
+		statement("struct OutputVert");
+	}
+	else
+	{
+		statement("struct OutputFrag");
+	}
+	begin_scope();
+	binding_number = 0;
+	variables.clear();
+	for (auto &id : ids)
+	{
+		if (id.get_type() == TypeVariable)
+		{
+			auto &var = id.get<SPIRVariable>();
+			auto &type = get<SPIRType>(var.basetype);
+
+			if (var.storage != StorageClassFunction && !var.remapped_variable && type.pointer &&
+			    var.storage == StorageClassOutput && interface_variable_exists_in_entry_point(var.self))
+			{
+				variables.push_back(&var);
+			}
+		}
+	}
+	sort(variables.begin(), variables.end(), VariableComparator(meta));
+	for (auto var : variables)
+	{
+		emit_interface_block_in_struct(*var, binding_number, false);
+	}
+	for (auto var : variables)
+	{
+		emit_interface_block_in_struct(*var, binding_number, true);
+	}
+	end_scope_decl();
+	statement("");
+
+	// Global variables.
+	for (auto global : global_variables)
+	{
+		auto &var = get<SPIRVariable>(global);
+		if (var.storage != StorageClassOutput)
+		{
+			add_resource_name(var.self);
+			statement("static ", variable_decl(var), ";");
+			emitted = true;
+		}
+	}
+
+	if (emitted)
+		statement("");
+
+	if (requires_op_fmod)
+	{
+		statement("float mod(float x, float y)");
+		begin_scope();
+		statement("return x - y * floor(x / y);");
+		end_scope();
+		statement("");
+	}
+}
+
+void CompilerHLSL::emit_buffer_block(const SPIRVariable &var)
+{
+	auto &type = get<SPIRType>(var.basetype);
+
+	add_resource_name(type.self);
+
+	statement("struct _", to_name(type.self));
+	begin_scope();
+
+	type.member_name_cache.clear();
+
+	uint32_t i = 0;
+	for (auto &member : type.member_types)
+	{
+		add_member_name(type, i);
+
+		auto &membertype = get<SPIRType>(member);
+		statement(member_decl(type, membertype, i), ";");
+		i++;
+	}
+	end_scope_decl();
+	statement("");
+
+	statement("cbuffer ", to_name(type.self));
+	begin_scope();
+	statement("_", to_name(type.self), " ", to_name(var.self), ";");
+	end_scope_decl();
+}
+
+void CompilerHLSL::emit_push_constant_block(const SPIRVariable &)
+{
+	statement("constant block");
```

Comment:
placeholder? push constant blocks should just be regular constant buffers I'd think, but you have to be careful about buffer layouts.